stan/math/rev/core/vari.hpp

```diff
@@ -179,11 +165,375 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
+   *  will be called in the reverse order of construction.
+   *
+   * @tparam S A sparse Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   * @param stacked If false will put this this vari on the nochain stack so
+   * that its `chain()` method is not called.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  vari_value(S&& x, bool stacked) : val_(x), adj_(x), chainable_alloc() {
+    this->set_zero_adjoint();
+    if (stacked) {
+      ChainableStack::instance_->var_stack_.push_back(this);
+    } else {
+      ChainableStack::instance_->var_nochain_stack_.push_back(this);
+    }
+  }
+
+  /**
+   * Return the number of rows for this class's `val_` member
+   */
+  const Eigen::Index rows() const { return val_.rows(); }
+  /**
+   * Return the number of columns for this class's `val_` member
+   */
+  const Eigen::Index cols() const { return val_.cols(); }
+  /**
+   * Return the size of this class's `val_` member
+   */
+  const Eigen::Index size() const { return val_.size(); }
+
+  void chain() {}
+  /**
+   * Initialize the adjoint for this (dependent) variable to 1.
+   * This operation is applied to the dependent variable before
+   * propagating derivatives, setting the derivative of the
+   * result with respect to itself to be 1.
+   */
+  inline void init_dependent() {
+    for (int k = 0; k < adj_.outerSize(); ++k) {
+      for (typename PlainObject::InnerIterator it(adj_, k); it; ++it) {
+        it.valueRef() = 1.0;
+      }
+    }
+  }
+
+  /**
+   * Set the adjoint value of this variable to 0.  This is used to
+   * reset adjoints before propagating derivatives again (for
+   * example in a Jacobian calculation).
+   */
+  inline void set_zero_adjoint() noexcept final {
+    for (int k = 0; k < adj_.outerSize(); ++k) {
+      for (typename PlainObject::InnerIterator it(adj_, k); it; ++it) {
+        it.valueRef() = 0.0;
+      }
+    }
+  }
+
+  /**
+   * Insertion operator for vari. Prints the current value and
+   * the adjoint value.
+   *
+   * @param os [in, out] ostream to modify
+   * @param v [in] vari object to print.
+   *
+   * @return The modified ostream.
+   */
+  friend std::ostream& operator<<(std::ostream& os, const vari_value<T>* v) {
+    return os << "val: \n" << v->val_ << " \nadj: \n" << v->adj_;
+  }
+
+  /**
+   * Allocate memory from the underlying memory pool.  This memory is
+   * is managed as a whole externally.
+   *
+   * Warning: Classes should not be allocated with this operator
+   * if they have non-trivial destructors.
+   *
+   * @param nbytes Number of bytes to allocate.
+   * @return Pointer to allocated bytes.
+   */
+  static inline void* operator new(size_t nbytes) {
+    return ChainableStack::instance_->memalloc_.alloc(nbytes);
+  }
+
+  /**
+   * Delete a pointer from the underlying memory pool.
+   *
+   * This no-op implementation enables a subclass to throw
```

Comment:
Are these gonna get marked final?

Reply:
`final` is just for virtual functions. Since these are not virtual they can't be marked final